### PR TITLE
Calculate local clipping rectangle per-primitive run

### PR DESCRIPTION
**Pull request by mrobinson**
Instead of using the hierarchical nature of the ClipScrollTree to calculate local clipping rectangles for primitives, do it when we are about to render a particular primitive run. This will allow us to handle custom clip chains while maintaining local clipping rectangle optimizations. It also allows us to greatly simplify a lot of shader code that used to deal with both clipping and scrolling nodes, as the local clipping rectangle is now bundled into the primitives local_clip_rect.
    
TransformOrOffset is the main struct that handles the efficient transformation of clipping rectangles between compatible coordinate systems. It will avoid doing expensive matrix operations as much as possible, which should make these calculations cheaper -- important now that they are done per-primitive run.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2200)

### CODE REVIEW

**Comment by mrobinson** [comment]
Gecko try in progress here: https://treeherder.mozilla.org/#/jobs?repo=try&revision=181a770af0c959460218df6c7d0a481ada2c70d1

**Comment by mrobinson** [comment]
Looks like there are two unexpected passes and one test which also seems to be crashing in other, unrelated try pushes. It's WebRender-related though, so I'll do some more investigation into that particular crash.

**Comment by mrobinson** [comment]
The Talos panic does not seem to be related to this PR and is being tracked here: https://bugzilla.mozilla.org/show_bug.cgi?id=1423203#c10

File: webrender/src/util.rs [hunk 1]
```diff
@@ -355,3 +368,84 @@ impl MaxRect for DeviceRect {
...
+/// when possible when dealing with non-perspective axis-aligned transformations.
+#[derive(Debug, Clone)]
+pub enum TransformOrOffset {
+    /// A simple offset, which can be used without doing any matrix math.
+    Offset(LayerVector2D),
+
+    /// A transformation with an inverse. If the inverse isn't present, this isn't a 2D
+    /// transformation, which means we need to fall back to using inverse_rect_footprint.
+    /// Since this operation is so expensive, we avoid it for the 2D case.
+    Transform {
```
Comment (kvark) [inline webrender/src/util.rs]: nice type system usage!

File: webrender/src/util.rs [hunk 2]
```diff
@@ -355,3 +368,84 @@ impl MaxRect for DeviceRect {
...
+        }
+    }
+
+    pub fn unapply(&self, rect: &LayerRect) -> LayerRect {
+        match *self {
+            TransformOrOffset::Offset(offset) => rect.translate(&-offset),
+            TransformOrOffset::Transform { inverse: Some(inverse), .. }  =>
+                inverse.transform_rect(&rect),
+            TransformOrOffset::Transform { transform, inverse: None } =>
+                transform.inverse_rect_footprint(rect),
```
Comment (kvark) [inline webrender/src/util.rs]: looks very clean, this one!

**Comment by glennw** [comment]
Reviewed 27 of 27 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[webrender/res/prim_shared.glsl, line 167 at r1](https://reviewable.io:443/reviews/servo/webrender/2200#-L-q3Ta7Wy5kEQOPaLPP-r1-167:-L-sxpURwAx6zUeL4e8E:b-4wcim1) ([raw file](https://github.com/servo/webrender/blob/b2103433784ed544cf469a2ea34233b922d18198/webrender/res/prim_shared.glsl#L167)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

looks like `reference_frame_relative_scroll_offset` and `scroll_offset` are left uninitialized?
</blockquote></details>

I think those fields are unused now, so can probably be removed from both the CPU and GPU versions of the ClipScrollNodeData.

---

*[webrender/src/prim_store.rs, line 1240 at r1](https://reviewable.io:443/reviews/servo/webrender/2200#-L-syglmgYb6D9v_eIYo:-L-syglmgYb6D9v_eIYp:b3ozhng) ([raw file](https://github.com/servo/webrender/blob/b2103433784ed544cf469a2ea34233b922d18198/webrender/src/prim_store.rs#L1240)):*
> ```Rust
>         prim_index: PrimitiveIndex,
>         prim_context: &PrimitiveContext,
>         local_clip_rect: LayerRect,
> ```

Should this be in the PrimitiveContext? Not sure...

---

*[webrender/src/prim_store.rs, line 1578 at r1](https://reviewable.io:443/reviews/servo/webrender/2200#-L-sykpXL17z3gKAZZ9E:-L-sykpXL17z3gKAZZ9F:b-8xwirr) ([raw file](https://github.com/servo/webrender/blob/b2103433784ed544cf469a2ea34233b922d18198/webrender/src/prim_store.rs#L1578)):*
> ```Rust
>                         coordinate_system_id: prim_coordinate_system_id,
>                     },
>                     local_clip_rect: LayerRect::zero(), // This should be unused.
> ```

Could you expand on that comment?

---

*[webrender/src/util.rs, line 372 at r1](https://reviewable.io:443/reviews/servo/webrender/2200#-L-swqLw7PKmFeCwz9FX:-L-swqLw7PKmFeCwz9FY:b8a9ykt) ([raw file](https://github.com/servo/webrender/blob/b2103433784ed544cf469a2ea34233b922d18198/webrender/src/util.rs#L372)):*
> ```Rust
> }
> 
> /// An enum that ensures tries to avoid expensive transformation matrix calculations
> ```

nit: grammar

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2200)*

File: webrender/res/prim_shared.glsl [hunk 3]
```diff
@@ -156,51 +155,20 @@ ClipScrollNode fetch_clip_scroll_node(int index) {
...
-    vec4 clip_rect = TEXEL_FETCH(sClipScrollNodes, uv0, 0, ivec2(4, 0));
-    node.local_clip_rect = clip_rect;
-
-    vec4 offsets = TEXEL_FETCH(sClipScrollNodes, uv0, 0, ivec2(5, 0));
-    node.reference_frame_relative_scroll_offset = offsets.xy;
-    node.scroll_offset = offsets.zw;
-
-    vec4 misc = TEXEL_FETCH(sClipScrollNodes, uv0, 0, ivec2(6, 0));
+    vec4 misc = TEXEL_FETCH(sClipScrollNodes, uv0, 0, ivec2(4, 0));
     node.is_axis_aligned = misc.x == 0.0;
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply webrender/res/prim_shared.glsl]: Removed these.

File: webrender/src/clip_scroll_node.rs [hunk 4]
```diff
@@ -122,9 +122,12 @@ pub struct ClipScrollNode {
     /// The axis-aligned coordinate system id of this node.
     pub coordinate_system_id: CoordinateSystemId,
 
+    pub coordinate_system_relative_transform: TransformOrOffset,
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply webrender/src/clip_scroll_node.rs]: I added a comment here.

File: webrender/src/clip_scroll_node.rs [hunk 5]
```diff
@@ -122,9 +122,12 @@ pub struct ClipScrollNode {
...
+
     /// A linear ID / index of this clip-scroll node. Used as a reference to
     /// pass to shaders, to allow them to fetch a given clip-scroll node.
     pub node_data_index: ClipScrollNodeIndex,
+
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply webrender/src/clip_scroll_node.rs]: Removed.

File: webrender/src/clip_scroll_node.rs [hunk 6]
```diff
@@ -187,10 +191,11 @@ impl ClipScrollNode {
     ) -> Self {
         let identity = LayoutTransform::identity();
         let info = ReferenceFrameInfo {
-            resolved_transform: LayerToScrollTransform::identity(),
+            resolved_transform: LayerTransform::identity(),
             source_transform: source_transform.unwrap_or(PropertyBinding::Value(identity)),
             source_perspective: source_perspective.unwrap_or(identity),
             origin_in_parent_reference_frame,
+            uninvertible: false,
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply webrender/src/clip_scroll_node.rs]: Changed this one.

**Comment by mrobinson** [comment]
Okay. Sorry for the delay on the new patch. After talking with Glenn about this, I finally have a new version which adds an entirely new texture for the local clips rects (now called "clip chain rects"). The primitive rectangle is calculated early by intersecting the local clip rect with the appropriate clip chain rect when the primitive is loaded in the shader. This ensures that that they clip chain rectangles are up to date even with property binding updates, but prevents a lot of churn in the GPU cache. The addition of the new texture is unfortunate, but this seemed like the lesser hack of all the things considered.

**Comment by mrobinson** [comment]
Here is the try run for this version: https://treeherder.mozilla.org/#/jobs?repo=try&revision=15c6957adf8edbd4d347c3bdbfe9665bd6633dcd

**Comment by kvark** [comment]
> this ensures that that they clip chain rectangles are up to date even with property binding updates, but prevents a lot of churn in the GPU cache.

I haven't reviewed the new code, but this sounds suspicious. We shouldn't be moving away from GPU cache back in the land of free-for-all textures just because some use cases are slow to update. I'd rather have us focused on getting the GPU cache performance solid, since it is simpler and has greater positive impact on the rest of the code. If the clip rect updates are slow, we may try the following things:
  - enable scattered GPU updates (in RendererOptios)
  - change the GPU cache allocation logic so that particular types (e.g. clip rectangles) or sub-types (e.g. animated clip rectangles) are grouped together in rows, so that the updates are efficient
  - play with row size
  - try sub-row updates according to simple heuristics
  - (possibly, other ways)

**Comment by glennw** [comment]
The issue it prevents in the GPU cache is that when we invalidate the item for a text run, that involves re-uploading the entire set of glyphs for a text run, which involves a lot of extra work (we found about an extra 0.5ms in compositor CPU time, but there is also the backend time in re-copying those glyphs across).

We *could* have an extra entry in the GPU cache, but that would involve adding an extra ```i32``` to all the instance structures, to make room for a GPU cache address. With this system, we can use a spare ```u16``` in each instance to index a fixed size texture.

It's not ideal, but I think it's probably the best solution for now. In the future, once we are able to compress the primitive instance structures further, it probably makes sense to remove this and use a normal GPU cache entry instead?

**Comment by glennw** [comment]
This looks good to me. The try shows two new passes. @kvark Anything you want to check over on this or is it ready to go?

**Comment by kvark** [comment]
:lgtm:

---

Reviewed 4 of 11 files at r2.
Review status: 21 of 28 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2200#-:-L0fGvayNLP9r3NYg4Jz:bnfp4nl)*

**Comment by glennw** [comment]
@mrobinson This just needs a rebase, then r=glennw,kvark.

**Comment by mrobinson** [comment]
Okay. I've rebased this PR, but it seems that the newly added test (reftests/boxshadow/overlap1.yaml) is failing with my PR, so I'll investigate this after Christmas.

**Comment by mrobinson** [comment]
Looks like the failing test is just a subpixel difference, so I've updated the png reference file and will push the rebased branch.

**Comment by staktrace** [comment]
@mrobinson This PR seems to have introduced or exposed a crash in WR, which manifests as the talos-g1 test failing. There's a backtrace from the log at https://bugzilla.mozilla.org/show_bug.cgi?id=1426116#c7

**Comment by mrobinson** [comment]
@staktrace I've posted a fix for this issue here: https://github.com/servo/webrender/pull/2249

Sorry for the delay, it was due to the holidays.

**Comment by staktrace** [comment]
Confirmed that the fix worked, thanks!

**Pull request merged**